core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -141,10 +142,21 @@ object Uri {
   }
 
   /** Decodes the String to a [[Uri]] using the RFC 3986 uri decoding specification */
-  def fromString(s: String): ParseResult[Uri] =
-    new RequestUriParser(s, StandardCharsets.UTF_8).Uri
-      .run()(PbParser.DeliveryScheme.Either)
-      .leftMap(e => ParseFailure("Invalid URI", e.format(s)))
+  def fromString(s: String): ParseResult[Uri] = {
+
+    val result: Try[ParseResult[Uri]] =
+      Try {
```

Comment:
```scala
Either.catchNonFatal{ block 
}.fold(Either.left[ParseFailure, Uri](_))(identity)
```

Reply:
Thanks for the review, @ChristopherDavenport. I'll make that change later today!

Aside - you find `Either#catchNonFatal` stylistically cleaner, hence your review comment? I'm just asking to explore the distinction, if any, between `Try#apply` and what you noted.

Reply:
It just brings in a new type for an equivalent type that is easier to unify. I find it easier, and cleaner in general.

Reply:
You probably want `Either.catchOnly[ParseFailure](...)` instead

Reply:
Why would it be better to catch only `ParseFailure`'s, i.e. propagate non-`ParseFailure`'s, @SystemFw, could you please say?

Reply:
Actually my bad, it depends on what you want to do. Just go with Chris suggestion and sorry for the confusion :)

Reply:
> Actually my bad

Not your bad at all! I was asking to learn (as I did talking to you at NEScala this year w/ Chris), @SystemFw, in case I was missing something.

> it depends on what you want to do.

It seemed worthwhile to me to return a `Left` for any `ParseFailure` due to a `NonFatal` being thrown. However, I've never touched this code before, so I admit I could be missing something. Please let me know if you disagree with `Either#catchNonFatal`.

Reply:
Could we just handle this directly in the Port rule of the parser?  An action function can return a rule.  Here's an example from the parboiled docs:

```scala
(foo: Rule1[Int]) ~> (i => test(i % 2 == 0) ~ push(i))
```

If we took that approach, we wouldn't have to worry about catching other funky exceptions in this much wider scope.

Reply:
> Could we just handle this directly in the Port rule of the parser?

That change seems more surgical and cleaner to me.

> If we took that approach, we wouldn't have to worry about catching other funky exceptions in this much wider scope.

If I understand your comment, you recommend I remove catching and handling `NonFatal`'s given that I'll make the change in the `Port`'s parser?

Reply:
I deleted my comment since I have a different question.

> Could we just handle this directly in the Port rule of the parser?

Could you please say more, @rossabaker? Looking at the `Port`'s code:

```scala
  def Port = rule {
    ":" ~ (capture(oneOrMore(Digit)) ~> { s: String =>
      (Some(s.toInt))
    } | push(None)) | push(None)
  }
```

are you suggesting that I ignore the `oneOrMoreDigit` if it does not evaluate to an `Int`, e.g. it's bigger than an `Int`'s max value?

Reply:
Yes, I think if you create  `test` rule, you can range check the port, and have the rule fail if it's out of bounds.

Reply:
Thanks, Ross. That helps. I'll make that change.